libsolidity/analysis/ReferencesResolver.cpp

```diff
@@ -301,109 +301,66 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 		using Location = VariableDeclaration::Location;
 		Location varLoc = _variable.referenceLocation();
 		DataLocation typeLoc = DataLocation::Memory;
-		// References are forced to calldata for external function parameters (not return)
-		// and memory for parameters (also return) of publicly visible functions.
-		// They default to memory for function parameters and storage for local variables.
-		// As an exception, "storage" is allowed for library functions.
+
 		if (auto ref = dynamic_cast<ReferenceType const*>(type.get()))
 		{
-			bool isPointer = true;
-			if (_variable.isExternalCallableParameter())
+			// If given location is not in allowed storage locations, throw error.
+			set<Location> allowedStorageLocations = _variable.allowedStorageLocations();
+			if (!allowedStorageLocations.count(varLoc))
 			{
-				auto const& contract = dynamic_cast<ContractDefinition const&>(
-					*dynamic_cast<Declaration const&>(*_variable.scope()).scope()
-				);
-				if (contract.isLibrary())
+				string allowedStorageLocationsStr;
+				string locStr;
+				for (auto const loc : allowedStorageLocations)
 				{
-					if (varLoc == Location::Memory)
-						fatalTypeError(_variable.location(),
-							"Location has to be calldata or storage for external "
-							"library functions (remove the \"memory\" keyword)."
-						);
+					if (!allowedStorageLocationsStr.empty())
+						allowedStorageLocationsStr += " or ";
+					if (loc == Location::Default)
+						locStr = "none";
+					else
+						locStr = "\"" + _variable.locationToString(loc) + "\"";
+					allowedStorageLocationsStr += locStr;
 				}
-				else
+				string errorString = "Storage location must be " + allowedStorageLocationsStr;
```

Comment:
I think the documentation mostly calls this `data location` and not `Storage location` - I also think `data location` is less confusing.